feat!: Add compositional experiments & metrics

vkakerbeck opened the pull request:
Adding configs & new evaluation measures for learning and recognizing compositional models.

I will rerun pre-training and inference on AWS and update the benchmark results table after the general review of this PR.

tristanls commented on src/tbp/monty/frameworks/utils/live_plotter.py:
```diff
@@ -147,7 +148,7 @@ def show_view_finder(
         if hasattr(first_learning_module, "current_mlh"):
             mlh = first_learning_module.get_current_mlh()
-            if mlh is not None:
+            if mlh is not None and mlh["graph_id"] != "no_observations_yet":
```
**issue**: Seems like instead of the `"no_observations_yet"` sentinel (I just now became aware of it), we should use `None`. This way, this extra check is not needed, and `self.current_mlh` is allowed to be `None`. As I read through the code, I don't think `mlh` can ever be `None`.

vkakerbeck replied:
Yeah that's a good point. I think the reason we set it to "no_observations_yet" (and also set defaults for the other current_mlh fields) is so we have all these fields available for logging, even when there were no observations (i.e. put "no_observations_yet" into the mlh column in .csv file is an LM didn't receive observations). But we could also set those defaults in the Logger instead and check for `current_mlh is None` there. I would do this in a separate PR though since it is not related to the compositional testbed (and might involve a bunch more unrelated changes).

tristanls commented on src/tbp/monty/frameworks/utils/live_plotter.py:
```diff
@@ -147,7 +148,7 @@ def show_view_finder(
             self.ax[0].add_patch(square)
         if hasattr(first_learning_module, "current_mlh"):
```
**issue**: Noticing now that this check isn't directly related to the line below. It should be `if hasattr(first_learning_module, "get_current_mlh"):`

tristanls reviewed (commented):
Sorry,  I didn't get a chance to review anything really. But I did stumble on `mlh` never being `None`. So I wanted to add these comments.

Perhaps a separate PR that changes `mlh` sentinel to be `None` instead of checking if "graph_id" property is set to "no_observations_yet" would be an improvement.

jeremyshoemaker commented on src/tbp/monty/frameworks/environments/embodied_data.py:
```diff
@@ -336,6 +347,16 @@ def change_object_by_idx(self, idx):
+                logger.warning(
+                    f"target object {self.primary_target['object']} not in",
+                    " parent_to_child_mapping",
+                )
```
Does it make sense to run this code in the event that `self.parent_to_child_mapping` is an empty dictionary, i.e. `{}`? Currently it will be run in that case, but if it doesn't make sense (I could see it either way), it would make the conditional simpler to just do `if self.parent_to_child_mapping` and that would exclude both `None` and `{}`.

jeremyshoemaker commented on src/tbp/monty/frameworks/experiments/pretraining_experiments.py:
```diff
@@ -139,16 +164,16 @@ def pre_episode(self):
+    def pre_epoch(self):
+        super().pre_epoch()
+
     def post_epoch(self):
```
Is there any reason why we're not calling `super().post_epoch()` somewhere in this method? Or are we completely overriding its behaviour?

jeremyshoemaker commented on src/tbp/monty/frameworks/loggers/graph_matching_loggers.py:
```diff
@@ -255,6 +260,17 @@ def update_overall_stats(self, mode, episode, episode_steps, monty_matching_step
+                int(performance == "consistent_child_obj")
+                or int(performance == "correct")
+                or int(performance == "correct_mlh")
+            )
```
This code seems too clever. Since these values are mutually exclusive, then the only possible values we're adding to this stat is 1 or 0. A better way to write this would be something like:
```suggestion
            if performance in {"consistent_child_obj", "correct", "correct_mlh"}:
                stats["num_correct_child_or_parent"] += 1
```

We could also use a list here by changing the `{}` to `[]`, but I think a set is faster for membership checks like this.

jeremyshoemaker commented on src/tbp/monty/frameworks/models/evidence_matching/hypotheses_updater.py:
```diff
@@ -219,6 +226,10 @@ def update_hypotheses(
             return []
 
         hypotheses_updates = []
+        telemetry: dict[str, Any] = {}
```
Is `Any` the most specific we can get for the values in this dictionary?

vkakerbeck replied:
I'm not sure, I just copied this from https://github.com/thousandbrainsproject/tbp.monty/blob/f0e181c4f8f66103d941df89f622db29aee840fc/src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py#L277

jeremyshoemaker commented on src/tbp/monty/frameworks/models/evidence_matching/learning_module.py:
```diff
@@ -445,18 +447,20 @@ def get_output(self):
             },
             non_morphological_features={
                 "object_id": object_id_features,
                 # TODO H: test if this makes sense to communicate
```
Should we address this TODO now? I'm not sure what it's suggesting.

jeremyshoemaker commented on src/tbp/monty/frameworks/models/evidence_matching/learning_module.py:
```diff
@@ -1183,6 +1189,46 @@ def _add_votes_to_buffer_stats(self, vote_data):
+                channel_prediction_error = displacer_telemetry.mlh_prediction_error
+                prediction_errors.append(channel_prediction_error)
+
+        if len(prediction_errors) > 0:
```
Another instance of checking whether a list contains items.
```suggestion
        if prediction_errors:
```

jeremyshoemaker commented on src/tbp/monty/frameworks/models/object_model.py:
```diff
@@ -534,7 +535,13 @@ def _initialize_and_fill_grid(
         # applied to all following locations.
-        self._initialize_location_mapping(start_location=locations[0])
+        # Sort out nan locations. For some reason the first location is sometimes nan.
+        # TODO - C: Why is there a nan location?
```
😄 Well, why is there a NaN location? Is this something we can easily fix now, or is it better to wait?

vkakerbeck replied:
I am not sure :D I thought I had figured it out but then it resurfaced I think.I think it has something to do with the way the buffer fills in nan values when a channel appears at a later step. I can take a closer look again

jeremyshoemaker commented on src/tbp/monty/frameworks/models/object_model.py:
```diff
@@ -794,20 +801,31 @@ def _get_new_voxel_features(
-            if self._observation_count[voxel[0], voxel[1], voxel[2], 0] > len(feats):
+            num_obs_in_voxel = self._observation_count[voxel[0], voxel[1], voxel[2], 0]
+            num_new_obs = len(feats)
+            if num_obs_in_voxel > num_new_obs:
```
I like that the two parts were separated out into named variables, as it makes this easier to understand.

jeremyshoemaker commented on src/tbp/monty/frameworks/utils/logging_utils.py:
```diff
@@ -809,6 +814,15 @@ def add_evidence_lm_episode_stats(lm, stats):
+    # was already an exact match with the label.
+    if (
+        stats["primary_performance"] in ["confused", "confused_mlh"]
+        and consistent_child_objects is not None
```
The argument `consistent_child_objects` seems to be a collection from it's use in this function. I think in that case it being None and it being empty should be handled the same and thus this can be simplified to:
```suggestion
        and consistent_child_objects
```

vkakerbeck commented on src/tbp/monty/frameworks/models/evidence_matching/learning_module.py:
```diff
@@ -1183,6 +1189,46 @@ def _add_votes_to_buffer_stats(self, vote_data):
+                    channel_telemetry.channel_hypothesis_displacer_telemetry
+                )
+                channel_prediction_error = displacer_telemetry.mlh_prediction_error
+                prediction_errors.append(channel_prediction_error)
```
makes sense, thanks

vkakerbeck commented on src/tbp/monty/frameworks/models/object_model.py:
```diff
@@ -794,20 +801,31 @@ def _get_new_voxel_features(
+                    if avg_feat != previous_average:
+                        if num_old_obs > num_new_obs:
+                            avg_feat = previous_average
+                        avg_feat = previous_average
```
nice catch! But I think we would want to keep the conditional and remove the line below it (i.e. if we have more new than old obs, we do want to use avg_feat and not overwrite it

vkakerbeck reviewed (commented):
Nice, thanks for all the feedback and suggestions!

jeremyshoemaker commented on src/tbp/monty/frameworks/config_utils/make_dataset_configs.py:
```diff
@@ -49,6 +59,7 @@ class ExperimentArgs:
     model_name_or_path: str = ""
     min_lms_match: int = 1
     seed: int = 42
+    supervised_lm_ids: Union[List[str], Literal["all"]] = field(default_factory=list)
```
This looks fine to me. I forgot about setting defaults in a dataclass like this. I was thinking of the case of setting a default in a function argument list and the problems with using list or dictionary literals in those places. This is a bit different, and using the factory approach sidesteps the issue I was mentioning in our huddle.

jeremyshoemaker commented on src/tbp/monty/frameworks/experiments/pretraining_experiments.py:
```diff
@@ -67,21 +67,11 @@ def run_episode(self):
-            lm.buffer.stats["detected_scale"] = target["scale"]
+        # Save compute if we are providing labels to all models, so don't need to
+        # perform matching parts of LM updates (default is matching_step)
+        if len(self.supervised_lm_ids) == len(self.model.learning_modules):
```
I think in most cases this will be fine, but we're not actually confirming that these two lists have the same elements. I think the quickest way to do this would be to turn them both into sets and compare the sets.
```suggestion
        if set(self.supervised_lm_ids) == set(self.model.learning_modules):
```

jeremyshoemaker commented on src/tbp/monty/frameworks/loggers/graph_matching_loggers.py:
```diff
@@ -255,6 +260,14 @@ def update_overall_stats(self, mode, episode, episode_steps, monty_matching_step
             stats["monty_steps"].append(episode_steps)
             stats["monty_matching_steps"].append(monty_matching_steps)
+            # older LMs don't have prediction error stats
+            if "episode_avg_prediction_error" in episode_stats.keys():
```
This shouldn't need the `keys()` call. Using `in` on a dictionary checks for whether the value is a key in the dictionary.
```suggestion
            if "episode_avg_prediction_error" in episode_stats:
```

jeremyshoemaker commented on src/tbp/monty/frameworks/models/evidence_matching/learning_module.py:
```diff
@@ -718,6 +724,7 @@ def _update_possible_matches(self, query):
         # Call this update in the step method?
         self.possible_matches = self._threshold_possible_matches()
+        self.previous_mlh = copy.deepcopy(self.current_mlh)
         self.current_mlh = self._calculate_most_likely_hypothesis()
```
Do we need the `deepcopy` here? On line 728, we're assigning a new value to `current_mlh`, so after these two lines (without the `deepcopy`) we'd have:

self.previous_mlh pointing to the old value of current_mlh
self.current_mlh pointing to the new value created inside _calculate_most_likely_hypothesis

I think the `deepcopy` call is just extra unnecessary work here.
```suggestion
        self.previous_mlh = self.current_mlh
        self.current_mlh = self._calculate_most_likely_hypothesis()
```

jeremyshoemaker commented on tests/unit/hierarchy_test.py:
```diff
@@ -0,0 +1,419 @@
+                do_train=False,
+                min_lms_match=1,
+                n_eval_epochs=2,
+                model_name_or_path=self.compositional_save_path + "/pretrained",
```
This should use `os.path.join` or the newer `pathlib` equivalents to avoid any path separator differences between platforms. I've put off suggesting changing everything over to `pathlib` since there's a lot of the older style in the codebase, so I'm fine with using the `os.path` approach.
```suggestion
                model_name_or_path=os.path.join(self.compositional_save_path, "pretrained"),
```

jeremyshoemaker reviewed (commented):
Not sure how much what I'm commenting on this time is new, or just stuff I missed the first time around. Looks good for the most part, with some minor things left.

jeremyshoemaker commented on src/tbp/monty/frameworks/environments/embodied_data.py:
```diff
@@ -336,6 +349,17 @@ def change_object_by_idx(self, idx):
+            self.consistent_child_objects = self.parent_to_child_mapping[
+                self.primary_target["object"]
+            ]
+        elif self.parent_to_child_mapping.keys():
```
Checking to see if a dictionary is empty is more idiomatically done this way:
```suggestion
        elif self.parent_to_child_mapping:
```
The comment also should be updated or removed. I'm not sure it's needed after this change.

jeremyshoemaker commented on src/tbp/monty/frameworks/models/evidence_matching/learning_module.py:
```diff
@@ -718,6 +723,7 @@ def _update_possible_matches(self, query):
         # Call this update in the step method?
         self.possible_matches = self._threshold_possible_matches()
+        self.previous_mlh = self.current_mlh
         self.current_mlh = self._calculate_most_likely_hypothesis()
```
This is more of an FYI, since I don't think it makes sense to do it here with such long variable and function names, but another way to write this that you'll see often in Python code is to do both assignments at the same time.
```suggestion
        self.previous_mlh, self.current_mlh = self.current_mlh, self._calculate_most_likely_hypothesis()
```

jeremyshoemaker reviewed (commented):
Fewer comments this time 😄

jeremyshoemaker reviewed (approved):
Unless there are other changes you need to make, this looks good to me.

vkakerbeck commented:
Okay, I ran all experiments + pretraining on AWS and updated the benchmark table. In the process of that I found another small bug in the GSG (only showed up on rare occasions so wasn't caught in the runs locally) and updated the wandb logging to work when run in parallel.

This should be all for the first version of the new benchmark. I made a list of next steps I can look into next: https://docs.google.com/document/d/1hN-R33GLnLKDR98PIXJH9nxyxhcBCABmiFn8Q7xjLew/edit?usp=sharing

jeremyshoemaker commented on src/tbp/monty/frameworks/loggers/graph_matching_loggers.py:
```diff
@@ -412,6 +412,10 @@ def get_formatted_overall_stats(self, mode, episode):
+            "episode/consistent_child_obj": stats["episode_consistent_child_obj"],
+            "episode/consistent_child_or_parent": stats["episode_consistent_child_obj"]
+            or stats["episode_correct"]
+            or stats["episode_correct_mlh"],
```
Does this work correctly? It looks wrong to me, usually because Python will assume the end of a statement if there isn't something to indicate that it continues onto the next line. Maybe it works because it's expecting a `,` but the formatting of this looks wrong to me. I'm surprised that `ruff format --check` is okay with this.

vkakerbeck replied:
autoformat broke it down like that to me. We are already doing this on some other lines further up but I can add some brackets just to be safe. Rerunning the experiment again to double check I get the same results.

vkakerbeck replied:
good to know! fwiw it doesn't seem to be any different functionally, I still get the exact same results

jeremyshoemaker reviewed (commented):
Found a typo, and also have a question about some code that looks weird to me.
